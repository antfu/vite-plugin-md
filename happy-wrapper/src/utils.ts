import { flow, pipe } from 'fp-ts/lib/function'
import { createDocument, createElement, createFragment, createTextNode } from './create'
import { HappyMishap } from './errors'
import type { Container, ContainerOrHtml, HTML, NodeSolver, NodeSolverReceiver, NodeType } from './happy-types'
import { isDocument, isElement, isElementLike, isFragment, isTextNode } from './type-guards'

/**
 * Determines the "content-type" of a given node
 */
export const getNodeType = (node: Container | HTML): NodeType => {
  if (typeof node === 'string')
    return 'html'

  return isTextNode(node)
    ? 'text'
    : isElement(node)
      ? 'element'
      : isDocument(node)
        ? 'document'
        : isFragment(node)
          ? 'fragment'
          : 'node'
}

/**
 * A helper utility to help convert DOM nodes or HTML to common type.
 *
 * Start by providing the _exclusions_ you want to make for input. By default, all
 * `Container` types are allowed along with `HTML`
 */
export const solveForNodeType: NodeSolver = (_ = undefined as never) => {
  const solver = <EE extends NodeType, OO>(): NodeSolverReceiver<EE, OO> => ({
    solver: s =>
      (node, parent) => {
        if (node === null)
          throw new Error('Value passed into solver was NULL!')
        if (node === undefined)
          throw new Error('Value passed into solver was UNDEFINED!')

        const type = getNodeType(node)
        if (type in s) {
          const fn = (s as any)[type]
          return fn(node, parent)
        }
        else {
          if (type === 'node' && 'element' in s && isElement(node)) {
            const fn = (s as any).element
            return fn(node, parent)
          }
          else if (type === 'node' && 'text' in s && isTextNode(node)) {
            const fn = (s as any).text
            return fn(node)
          }

          throw new HappyMishap(`Problem finding "${type}" in solver.`, { name: `solveForNodeType(${type})` })
        }
      }
    ,
  })
  return {
    outputType: () => solver(),
    mirror: () => solver(),
  }
}

/**
 * Ensures any Container, array of Containers, or even HTML or HTML[] are all
 * normalized down to just HTML.
 */
export function toHtml<D extends ContainerOrHtml | ContainerOrHtml[] | null>(node: D): HTML {
  if (node === null)
    return ''

  const n = (!Array.isArray(node) ? [node] : node) as ContainerOrHtml[]
  try {
    const results = n.map((i) => {
      const convert = solveForNodeType()
        .outputType<HTML>()
        .solver({
          html: h => h,
          text: t => t.textContent,
          element: e => e.outerHTML,
          node: (ne) => {
            if (isElement(ne))
              convert(ne)
            if (isTextNode(ne))
              convert(ne)

            throw new Error(
              `Unknown node type detected while converting to HTML: [ name: ${ne.nodeName}, type: ${ne.nodeType}, value: ${ne.nodeValue} ]`,
            )
          },
          document: d => `<html>${d.head.hasChildNodes() ? d.head.outerHTML : ''}${d.body.outerHTML}</html>`,
          fragment: (f) => {
            if (isElementLike(f))
              return f.firstElementChild.outerHTML

            else
              return f.childNodes.map(c => convert(c, f)).join('')
          },
        })

      return convert(i)
    })

    return results.join('')
  }
  catch (e) {
    if (Array.isArray(node))
      throw new HappyMishap(`Problem converting an array of ${n.length} nodes [${n.map(i => getNodeType(i as any)).join(', ')}] to HTML`, { name: 'toHTML([...])', inspect: ['first node', node[0]], error: e })

    else
      throw new HappyMishap(`Problem converting "${getNodeType(node)}" to HTML!`, { name: 'toHTML(getNodeType(node))', inspect: node, error: e })
  }
}

/**
 * Clones most DOM types
 */
export function clone<T extends Container | HTML>(container: T): T {
  const clone = solveForNodeType()
    .mirror()
    .solver({
      html: h => `${h}`,
      fragment: flow(toHtml, createFragment),
      document: (d) => {
        return createDocument(d.body.innerHTML, d.head.innerHTML)
      },
      element: e => pipe(e, toHtml, createElement),
      node: flow(toHtml, createFragment, f => f.firstElementChild ? f.firstElementChild : f.firstChild),
      text: flow(toHtml, createTextNode),
    })

  return clone(container)
}

/**
 * ensures that a given string doesn't have any HTML inside of it
 */
export function safeString(str: string): string {
  const node = createFragment(str)
  return node.textContent
}
