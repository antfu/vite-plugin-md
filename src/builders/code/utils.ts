
import type { CodeFenceMeta, CodeOptions, LineCallback, PrismOptions, ShikiOptions } from './types'
import { Highlighter } from './types'

export function isValidLanguage(h: Highlighter, _meta: CodeFenceMeta): boolean {
  if (h === Highlighter.prism) {
    //
  }
  else {
    //
  }
}

/** Type Guard to detect options hash for Prism */
export function usesPrismHighlighting(h: CodeOptions): h is PrismOptions {
  return h.engine === Highlighter.prism
}

/** Type Guard to detect options hash for Shiki */
export function usesShikiHighlighting(h: CodeOptions): h is ShikiOptions {
  return h.engine === Highlighter.shiki
}

export function isLineCallback(cb?: false | string | LineCallback): cb is LineCallback {
  return !!(cb && typeof cb === 'function')
}
