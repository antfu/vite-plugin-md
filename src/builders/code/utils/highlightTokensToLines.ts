import type { CodeBlockMeta } from '../code-types'

/** converts HighlightTokens to lines of code */
export function highlightTokensToLines(fence: CodeBlockMeta<'dom'>): number[] {
  const lines: number[] = []

  fence.highlightTokens.forEach((t) => {
    switch (t.kind) {
      case 'line':
        lines.push(t.line)
        break
      case 'range':
        {
          let i = t.from
          while (i <= t.to) {
            lines.push(i)
            i++
          }
        }
        break
      case 'symbol':
        // TODO: need to implement this
    }
  })

  return lines
}
