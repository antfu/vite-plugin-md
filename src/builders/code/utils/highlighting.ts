import type { CodeBlockMeta, CodeOptions, LineCallback, PrismOptions, ShikiOptions } from '../types'
import { Highlighter } from '../types'
import { validPrismLanguage } from './prism-languages'

export function isValidLanguage(h: Highlighter, meta: CodeBlockMeta<'dom'>): boolean {
  if (h === Highlighter.prism) {
    return validPrismLanguage(meta.lang)
  }
  else {
    // TODO: do for Shiki
    return true
  }
}

/** Type Guard to detect options hash for Prism */
export function usesPrismHighlighting(h: CodeOptions): h is PrismOptions {
  return h.engine === Highlighter.prism
}

/** Type Guard to detect options hash for Shiki */
export function usesShikiHighlighting(h: CodeOptions): h is ShikiOptions {
  return h.engine === Highlighter.shiki
}

export function isLineCallback(cb?: false | string | LineCallback): cb is LineCallback {
  return !!(cb && typeof cb === 'function')
}
