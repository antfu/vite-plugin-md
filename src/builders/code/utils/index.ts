export * from './prism-languages'
export * from './highlighting'
export * from './trace'
export * from './log'
