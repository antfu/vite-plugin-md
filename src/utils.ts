export function getVueVersion(defaultVersion = '3.2.0') {
  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    let v = require('vue')
    if (v.default)
      v = v.default
    return v.version || defaultVersion
  }
  catch (e) {
    return defaultVersion
  }
}
