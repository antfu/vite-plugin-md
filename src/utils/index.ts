export * from './core'
export * from './fp'
export * from './createFnWithProps'
export * from './valueOrCallback'
