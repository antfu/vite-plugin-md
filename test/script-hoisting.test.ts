import { describe, expect, it } from 'vitest'
import { composeFixture } from './utils'

describe('hoisted script blocks', () => {
  it('all "script setup" blocks merged into one block with frontmatter defined first', async () => {
    const { scriptSetup } = await composeFixture('hoisted-2', {
      headEnabled: true,
    })

    // ordering
    const imports = scriptSetup.indexOf('import ')
    const title = scriptSetup.indexOf('const title')
    const defineExpose = scriptSetup.indexOf('defineExpose')
    const sayHi = scriptSetup.indexOf('const sayHi')

    // imports should be first
    expect(title).toBeGreaterThan(imports)
    // then the frontmatter props, followed by defineExpose
    expect(defineExpose).toBeGreaterThan(title)
    // finally, non-import user blocks are injected
    expect(sayHi).toBeGreaterThan(defineExpose)
  })

  it('userland script blocks defined as separate blocks and after script setup', async () => {
    const { scriptSetup, scriptBlocks, component } = await composeFixture('hoisted-2')

    expect(scriptSetup).toContain('const sayHi')
    expect(scriptSetup).toContain('const test')

    const script = scriptBlocks.join('\n')

    expect(script).toContain('const sayBye')
    expect(script).toContain('const test1')
    expect(script).toContain('const test2')
    expect(script).not.toContain('const test =')

    const sayHi = component.indexOf('const sayHi')
    const sayBye = component.indexOf('const sayBye')
    expect(sayHi).lessThan(sayBye)
  })
})
