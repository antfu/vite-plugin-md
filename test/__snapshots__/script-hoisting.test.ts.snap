// Vitest Snapshot v1

exports[`hoisted script blocks > snapshot of component section remains the same 1`] = `
"<route lang=\\"json\\">{}</route>

<template>
  <div class=\\"markdown-body\\">
<h1>{{title}}</h1>
<p>The import above should be brought into the VueJS component which would then allow us to use this <em>fake</em> component in our page:</p>
<foobar title=\\"i am not real\\"></foobar>
<p>Furthermore we can add a script setup block like this:</p>

<p>And now the function we’ve defined can be used: {{ sayHi(‘Foo’) }}</p>
<p>This doesn’t just apply to <code class=\\"\\">&lt;script setup&gt;</code> blocks but also more traditional <code class=\\"\\">&lt;script&gt;</code> blocks too:</p>

<p>The traditional <em>script blocks</em> can appear multiple times, and even use different languages without incident. These blocks will be maintained as separate blocks in the resulting SFC component.</p>

</div>
</template>

<script setup>
  import { foobar } from 'some-other-place';
  defineExpose({ frontmatter: {\\"title\\":\\"Hoisted Scripts\\",\\"description\\":\\"when an import or script block is embedded in HTML we need to preserve this\\"} })
  
 const title = \\"Hoisted Scripts\\"
 const description = \\"when an import or script block is embedded in HTML we need to preserve this\\"
 const excerpt = undefined
  const sayHi = (name: string) => \`hi \${name}\`
</script>

<script>
  /** frontmatter meta-data for MD page **/
  export const frontmatter = {\\"title\\":\\"Hoisted Scripts\\",\\"description\\":\\"when an import or script block is embedded in HTML we need to preserve this\\"}
</script>
<script lang=\\"ts\\">
// note that this variable is available within the script block but _not yet_ the template
export const meToo = (thing: string) => \`I'm not materialistic but I want a \${thing} too!\`
// with a traditional script block we need to return the variable to expose it to the template
return { meToo }
</script>
<script lang=\\"js\\">
export const ANSWER = 42
</script>
"
`;

exports[`hoisted script blocks > snapshot of hoisted scripts section remains the same 1`] = `
[
  "<script lang=\\"ts\\" setup=\\"\\">
import { foobar } from 'some-other-place';
</script>",
  "<script lang=\\"ts\\" setup=\\"\\">
const sayHi = (name: string) => \`hi \${name}\`
</script>",
  "<script lang=\\"ts\\">
// note that this variable is available within the script block but _not yet_ the template
export const meToo = (thing: string) => \`I'm not materialistic but I want a \${thing} too!\`
// with a traditional script block we need to return the variable to expose it to the template
return { meToo }
</script>",
  "<script lang=\\"js\\">
export const ANSWER = 42
</script>",
]
`;
